### DISCUSSION TITLE
How to use "exec" with KubernetesClient.Classic?

### DISCUSSION THREAD

**Discussion by cody-malcolm** [kubernetes-client/csharp#887] [Q&A] [answered]
Hello, I'm having trouble getting an exec method working with the Classic version of KubernetesClient. I'm using the Classic version because we are using .NET Framework 4.7.2 (and can't upgrade) and also OIDC authentication. I wasn't able to get the standard version working at all with OIDC (using version 4.0.x,, the last version that supports .NET Framework >5).

When I switched to the Classic version, our authentication and authorization issues were completely resolved, and we're able to proof of concept every function we are looking for except for exec. I'm finding the following issues:

The method in the example (eg. `await client.WebSocketNamespacedPodExecAsync(pod.Metadata.Name, "default", "ls",
                    pod.Spec.Containers[0].Name).ConfigureAwait(false);`) does not work because the method doesn't exist. When checking the methods that are available, I found `client.CoreV1.ConnectGetNamespacedPodExecAsync` as well as a post version. But when using these methods, I get the following exception message: `k8s.Autorest.HttpOperationException: Operation returned an invalid status code 'BadRequest'
   at k8s.Kubernetes.<SendRequestRaw>d__33.MoveNext() in /_/src/KubernetesClient/Kubernetes.cs:line 127`.. I haven't been able to find much on the differences between the Classic and standard libraries, or how to use exec with the Classic library (assuming this is supposed to be possible). I know that exec should be able to work with this version of .NET Framework because I had exec working with a token and version 4.0.x of the standard library.

Are there more resources somewhere that I'm not finding? Is exec not supported with Classic? If not, could it be added (and is there reference material to help with doing so)?

A sample of the code I'm trying to use is below:

```        
private async static Task ExecInPod(IKubernetes client, k8s.Models.V1Pod pod)
        {
            var temp = await client.CoreV1.ConnectGetNamespacedPodExecAsync(pod.Metadata.Name, "default", "ls", pod.Spec.Containers[0].Name).ConfigureAwait(false);
 
        }

public async Task My_Test()
        {
            try
            {
                var config = KubernetesClientConfiguration.BuildConfigFromConfigFile("C:\\Code\\config");

                var client = new Kubernetes(config);

                Console.WriteLine("Starting Request!");
                                                             
                var podList = client.CoreV1.ListNamespacedPod("default");

                Console.WriteLine("Got the lists!");
                
                var pod = podList.Items[0];
                await ExecInPod(client, pod).ConfigureAwait(false);
                
            }
            catch (Exception e)
            {
                Console.WriteLine(e);
            }

        }
```
Thanks in advance.

**Comment by tg123** [comment, marked as answer]
`ConnectGetNamespacedPodExecAsync` actually return a websocket, the client side should `upgrade the connection`, that is why you see a badrequest I guess

should use `WebSocketNamespacedPodExecAsync` instead. However, it is not available in net47 because of lacking of websocket APIs.
I remember the webocket related APIs were first introduced in netstandard2.1 :(

**Reply by tg123** [reply]
The error is confused and the API should not be set `internal`

**Reply by cody-malcolm** [reply]
Thanks for the response. When I uninstall KubernetesClient.Classic v8 and install KubernetesClient v3.0.16 (which targets only netstandard2.0), the `WebSocketNamespacedPodExecAsync` method is available - isn't this contradictory? In any case, what is the bottom line? Is there any way to use exec functionality with this library with net47/netstandard2.0? I'm sorry if this is answered by your second reply, but I don't know what that means.

**Reply by tg123** [reply]
maybe i did not remember the time correctly.
did you test if the 3.0.6 websocket working?

if so, we could port it back to current version

**Reply by tg123** [reply]
check link here 
https://github.com/kubernetes-client/csharp/blob/0276bcc7085536c3569e09c3ef3e5a94f5cebca3/src/KubernetesClient/Kubernetes.WebSocket.cs#L322

this is something can be done in netstandard2.1+ only
I believe there would be some cert related issue if < netstandard2.1

**Reply by cody-malcolm** [reply]
>did you test if the 3.0.6 websocket working?
>
>if so, we could port it back to current version

I'm not too familiar with WebSockets but it appears to be working perfectly fine in version 4.0.26. The example code worked with `ls`, and I was also able to (for example) change the `ls` command to `new string[] { "touch", "testing.txt }`, and the file appears as expected. This is with net47.

**Reply by tg123** [reply]
thanks for confirming,
created a new work item to track

**Reply by cody-malcolm** [reply]
Thanks!
